<font color='green'> 
**Udemy - (50 Saat) Python A-Z™: Veri Bilimi ve Machine Learning**
10. Bölüm - Makine Öğrenmesi II - Doğrusal Regresyon ve Kuzenleri (265-267. videolar) - Vahit Keskin anlattı. </font>

# Multiple Linear Regression

### Özet 

**Data Loading**

**Data Preprocessing**
* X ve y değişkenlerini belirliyoruz.
* Scikit-learn ile data setini train ve test set olarak ayırıyoruz.

**Creating Model with Statsmodels**
* Modeli fit ettik.
* Modeli yorumluyoruz.
* `model.summary()`nin verdiği her tabloyu ayrı ayrı alabiliriz.
* R-squared, Adj. R-squared (düzeltilmiş R^2), F-statistic bunlar arasında ne fark vardır?

**Creating Model with Scikit-Learn**

**Tahmin**
* Tek bir gözlemle satış tahmini yapıyoruz.
* Modelimizin tahmin başarısı nedir?

**Model Tuning / Model Doğrulama**
* Modeli tekrar kurup RMSE ve R^2 değerlerini alıyoruz.
* Train seti üzerinden cross validation işlemi gerçekleştiriyoruz ve RMSE, R^2 skorlarını alıyoruz.
* Test seti üzerinden cross validation işlemi gerçekleştiriyoruz ve RMSE, R^2 skorlarını alıyoruz.

## Data Loading

In [1]:
import pandas as pd

ad = pd.read_csv("Advertising.csv", usecols = [1,2,3,4]) # ilk column Unnamed:0 geldiği için böyle yaptık.
df = ad.copy()
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [2]:
df.shape

(200, 4)

## Data Preprocessing

#### <font color='green'>X ve y değişkenlerini belirliyoruz.</font>

In [3]:
X = df.drop("sales", axis = 1)
y = df["sales"]

#### <font color='green'>Scikit-learn ile data setini train ve test set olarak ayırıyoruz.</font>

In [4]:
from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

# her train, test seti oluşturacağım zaman farklı değerler çıkmasın diye random_state=42 yazdık. 

In [5]:
X_train.shape

(160, 3)

In [6]:
y_train.shape

(160,)

In [7]:
X_test.shape

(40, 3)

In [8]:
y_test.shape

(40,)

## Creating Model with Statsmodels

<font color='blue'>Eğer yorumlama gibi ihtiyaçlarımız, beklentilerimiz varsa bu durumda "Statsmodels" kütüphanesini kullanmamız gerekiyor. Sadece diğer fonksiyonlarla da uyumlu, diğer makine öğrenmesi algoritmalarıyla da aynı dili konuşabileyim amacıyla bir işlem yapayım dersek scikit-learni kullanacağız. </font>

In [10]:
import statsmodels.api as sm
lm = sm.OLS(y_train, X_train)

In [11]:
model = lm.fit()

#### <font color='green'>Modeli yorumluyoruz.</font>

In [12]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              2935.
Date:                Sat, 18 Dec 2021   Prob (F-statistic):                   1.28e-137
Time:                        13:57:08   Log-Likelihood:                         -336.65
No. Observations:                 160   AIC:                                      679.3
Df Residuals:                     157   BIC:                                      688.5
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0531      0.001     36.467      0.000       0.050       0.056
radio          0.2188      0.011     20.138      0.000       0.197       0.240
newspaper      0.0239      0.008      3.011      0.003       0.008       0.040
==============================================================================
Omnibus:                       11.405   Durbin-Watson:                   1.895
Prob(Omnibus):                  0.003   Jarque-Bera (JB):               15.574
Skew:                          -0.432   Prob(JB):                     0.000415
Kurtosis:                       4.261   Cond. No.                         13.5
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<font color='blue'>
    
- R^2 değeri artmış çünkü modelde hem değişken artması anlamında değişkenler arttı (simple regression'da sadece TV değişkeni ile regression kurmuştuk) hem de gerçekten de açıklanabilirliği olan değişkenler var bu modelde. Televizyon harcamaları, radyo harcamaları ve gazete harcamaları şeklinde. Dolayısıyla R^2 değeri artmış. Satış değişkeninin değişkenliğinin yaklaşık %98'ini açıklıyoruz ki bu gerçek hayat için çok çok yüksek bir oran. 

- Modelimiz anlamlı F-statistic ve Prob(F-statistic) değerine baktığımızda. 

- Katsayıları da incelediğimizde bütün katsayılar anlamlı.

````
Yani bu modelde bağımlı değişken satışları açıklamak için modele aldığımız bütün değişkenler 
anlamlı. Kişisel tavsiyem: hangi algoritmayı kullanacak olursanız olun eğer hedefinizde bir 
sürekli değişken söz konusuysa öncelikle mutlaka bir doğrusal modele sokun. Bu değişkenlerin 
analitik anlamda bağımlı değişkene olan etkilerinin anlamlı olup olmadığını inceleyebilmek 
adına yapmalısınız. 

````
- Tamam değişkenlerin hepsinin anlamlı olduğunu anladık. Katsayıları yorumlayalım. (Burada 1 değerini eklemedik, sabit katsayı yok burada, istersek bunu ekleyebiliriz.) TV harcamalarındaki bir birimlik artış satışlarda ortalama 0.0531 artışa sebep olacaktır. Radyo harcamalarındaki bir birimlik artış satışlarda ortalama 0.2188 artışa sebep olacaktır. Gazete harcamalarındaki bir birimlik artış satışlarda ortalama 0.0239 artışa sebep olacaktır. 
 </font>

#### <font color='green'>`model.summary()`nin verdiği her tabloyu ayrı ayrı alabiliriz.</font>

In [18]:
model.summary().tables

[<class 'statsmodels.iolib.table.SimpleTable'>,
 <class 'statsmodels.iolib.table.SimpleTable'>,
 <class 'statsmodels.iolib.table.SimpleTable'>]

In [19]:
model.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

#### <font color='green'>R-squared, Adj. R-squared (düzeltilmiş R^2), F-statistic bunlar arasında ne fark vardır?</font>

<font color='blue'>
    
* Varyansın açıklanabilirliği: R^2 ve düzeltilmiş R^2
    
* Modelin anlamlılığı: F-statistic
    
* Tahmin başarısı: MSE, RMSE
    
R^2 ve düzeltilmiş R^2, bağımlı değişkendeki değişikliğin bağımsız değişkenlerce açıklanabilme oranını vermektedir. Yani modeldeki bağımsız değişkenler elimizdeki bağımlı değişkenin varyansının şu kadarını açıklar derler. F-istatistiği ise modelin anlamlılığını değerlendirir, model anlamlı mı değil mi? 

Bir de tahmin başarısı diye bir şey var. Doğrusal modellerde modellerin tahmin başarısını değerlendirmek için MSE, RMSE gibi değerlere bakıyor olacağız. 
</font>

## Creating Model with Scikit-Learn

In [13]:
from sklearn.linear_model import LinearRegression

lm = LinearRegression()
model = lm.fit(X_train, y_train)

In [14]:
model.intercept_ 

2.979067338122629

In [15]:
model.coef_

array([0.04472952, 0.18919505, 0.00276111])

## Tahmin

Model denklemi:

Sales = 2.97 + TV\*0.04 + radio\*0.18 + newspaper*0.002

Örneğin 30 birim TV harcaması, 10 birim radio harcamasi, 40 birimde gazete harcaması olduğunda satışların tahmini değeri ne olur?

#### <font color='green'>Tek bir gözlemle satış tahmini yapıyoruz. </font>

<font color='blue'>Elimize yeni veriler geldi. </font>

In [16]:
yeni_veri = [[30],[10],[40]]

<font color='blue'>Bunları dataframe'e çevirip transpozunu almamız gerekiyor.</font>

In [30]:
# deneme yaptım

e = [[30,20],[10,50],[40,60]]
pd.DataFrame(e).T

0   1   2
0  30  10  40
1  20  50  60

In [17]:
# pandas dataframe'ine çevirip transpozunu alıyoruz.

yeni_veri = pd.DataFrame(yeni_veri).T 

In [18]:
yeni_veri

0   1   2
0  30  10  40

In [19]:
model.predict(yeni_veri)

array([6.32334798])

#### <font color='green'>Modelimizin tahmin başarısı nedir?</font>

In [20]:
import numpy as np 
from sklearn.metrics import mean_squared_error

<font color='blue'>Eğitim hatasına erişiyoruz.</font>

In [21]:
rmse_train = np.sqrt(mean_squared_error(y_train, model.predict(X_train))) # root-mean-square error
rmse_train

1.644727765644337

<font color='blue'>Test hatasına erişiyoruz.</font>

In [22]:
rmse_test = np.sqrt(mean_squared_error(y_test, model.predict(X_test))) 
rmse_test

1.7815996615334502

## Model Tuning / Model Doğrulama

<font color='blue'>
Çoklu doğrusal regresyon modelinde β0'dan başka dışsal parametre olarak görülebilecek bir parametre yok, o da aslında dışsal parametre değil. Dolayısıyla burada model tuning işlemini model doğrulama işlemi olarak ele alacağız. Bizim model doğrulama yöntemleriyle yani test, train veya k katlı cross validation yöntemi gibi yöntemlerle yapmak istediğimiz şey elde ettiğimiz hataları daha doğru hatalar olarak elde etmeye çalışmak. 
</font>

### <font color='purple'>Modeli tekrar kurup RMSE ve R^2 değerlerini alıyoruz. </font>

#### <font color='green'>`random_state=42` yaptık.</font>

In [24]:
# yukarıdaki işlemleri tekrar yaptık. Modeli scikit-learn ile oluşturduk.
X = df.drop('sales', axis=1)
y = df["sales"]
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.20, 
                                                    random_state=42)
lm = LinearRegression() 
model = lm.fit(X_train, y_train)

In [25]:
# eğitim hatasına bakıyoruz (yukarıda da bakmıştık zaten -> rmse_train değeri.)
np.sqrt(mean_squared_error(y_train, model.predict(X_train))) 

1.644727765644337

In [26]:
# R^2 değerine bakacağız.

model.score(X_train, y_train)

0.8957008271017818

#### <font color='green'>`random_state=99` yaptık.</font>

<font color='blue'>`random_state` değerini değiştirdiğimizde RMSE ve R^2 değerleri de değişiyor.</font>

In [27]:
X = df.drop('sales', axis=1)
y = df["sales"]
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.20, 
                                                    random_state=99)
lm = LinearRegression() 
model = lm.fit(X_train, y_train)

In [28]:
np.sqrt(mean_squared_error(y_train, model.predict(X_train))) 

1.7236824822650751

In [29]:
model.score(X_train, y_train)

0.8906288862925659

#### <font color='green'>`random_state=144` yaptık.</font>

In [30]:
X = df.drop('sales', axis=1)
y = df["sales"]
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.20, 
                                                    random_state=144)
lm = LinearRegression() 
model = lm.fit(X_train, y_train)

In [31]:
np.sqrt(mean_squared_error(y_train, model.predict(X_train))) 

1.6748559274650712

In [32]:
model.score(X_train, y_train)

0.8971614078663419

<font color='blue'>`random_state`i değiştirdiğimizde data setinin farklı kısımlarından train test ayrımı yapıyor model. Ve seçilen veri değiştikçe RMSE ve R^2 değerleri de değişiyor. Model doğrulama için gerçekleştirdiğimiz **cross validation** işlemi datasetinin hangi kısmını seçeceğiz problemini ortadan kaldırıyor.   </font>

### <font color='purple'> Train seti üzerinden cross validation işlemi gerçekleştiriyoruz ve RMSE, R^2 skorlarını alıyoruz.</font>

##### <font color='green'>Train seti üzerinden 10 fold cross validation ile R^2 skorları alıyoruz.</font>

In [33]:
cross_val_score(model, X_train, y_train, cv=10, scoring = "r2")

array([0.86580772, 0.92273806, 0.90690214, 0.9124702 , 0.69765146,
       0.92654282, 0.92526348, 0.94823778, 0.92941476, 0.69875488])

##### <font color='green'>10 tane birbirinden farklı R^2 değeri geldi. Ortalamasını aldığımızda çıkan değer modelimizin daha güvenilir, doğrulanmış R^2 değeridir. </font>

In [34]:
cross_val_score(model, X_train, y_train, cv=10, scoring = "r2").mean()

0.8733783298422942

##### <font color='green'>Train seti üzerinden 10 fold CV yaparak RMSE skoru alıyoruz.</font>

In [35]:
cross_val_score(model, X_train, y_train, cv=10, scoring = "neg_mean_squared_error")

array([-3.77011117, -1.38904597, -1.31506551, -3.32109589, -8.82506973,
       -2.37926645, -2.2872061 , -1.05714426, -2.99532621, -3.15248307])

In [36]:
# negatif mean squared errorden üretilen sonuçlar negatif olarak üretildiği için başına - koyuyoruz.

-cross_val_score(model, X_train, y_train, cv=10, scoring = "neg_mean_squared_error")

array([3.77011117, 1.38904597, 1.31506551, 3.32109589, 8.82506973,
       2.37926645, 2.2872061 , 1.05714426, 2.99532621, 3.15248307])

In [37]:
# rmse değerini almak için yukarıdaki koda kök alma işlemi ekliyoruz.

np.sqrt(-cross_val_score(model, X_train, y_train, cv=10, scoring = "neg_mean_squared_error"))

array([1.94167741, 1.17857794, 1.14676306, 1.82238742, 2.97070189,
       1.5424871 , 1.51235118, 1.02817521, 1.73070108, 1.77552332])

<font color='blue'>Şu anda eğitim setimiz için 10 adet eğitim hatası oluşturuldu ve bu eğitim hatalarının her birisi birbirinden farklı. Bizim eğitim seti üzerinden `np.sqrt(mean_squared_error(y_train, model.predict(X_train)))` koduyla bulduğumuz rmse değeri 1.67 idi. Fakat verinin farklı bir parçasında 1.97, farklı bir parçasında 2.97, 1.77 gibi farklı farklı birçok değer var. Bu değerlerin ortalamasını aldığımızda bunu valide etmiş oluyorum ve diyorum ki benim gerçek eğitim hatam bu.  </font>

In [38]:
np.sqrt(-cross_val_score(model, X_train, y_train, cv=10, scoring = "neg_mean_squared_error")).mean()

1.6649345607872932

### <font color='purple'> Test seti üzerinden cross validation işlemi gerçekleştiriyoruz ve RMSE, R^2 skorlarını alıyoruz.</font>

#### <font color='green'>Cross validation yapmadan önce test seti üzerinden RMSE değeri alıyoruz.</font>

In [39]:
np.sqrt(mean_squared_error(y_test, model.predict(X_test))) 

1.6640263686701036

#### <font color='green'>Cross validation yaparak test setinin RMSE değerini alıyoruz.</font>

In [40]:
np.sqrt(-cross_val_score(model, X_test, y_test, cv=10, scoring = "neg_mean_squared_error")).mean()

1.7399924960346644

<font color='blue'>
Burada iki düzeyli dikkat edilmesi gereken konu var:
    
- Birincisi; train ve test hatasının birbirinden farklılaşıyor olacağı ve eğitim hatasının test hatasının kötü bir tahmincisi olması durumu. Çünkü baktığımızda eğitim hatasını 1.67, test hatasını 1.66 bulmuştuk (yine de iyi bir tahmin). 

- İkincisi; Elde ettiğimiz valide edilmemiş hatanın da yanıltıcı olduğunu görüyoruz. Baktığımızda train hatalarını karşılaştırdığımızda cross validation yapmadığımızda hatayı 1.67 bulmuşken valide ettiğimizde 1.66 bulduk. Valide edilmemiş test hatasını 1.66 bulmuştuk, halbuki gerçek test hatamız (valide edilmiş) 1.73. </font>